flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,449 @@
+      let capability = self.inner.createCountCap()
+      self.issued[capability.capabilityID] = true
+      return capability
+   }
+
+   fun revokeCountCap(capabilityID: UInt64) {
+      if self.issued.containsKey(capabilityID) {
+         self.issued.remove(key: capabilityID)
+         self.inner.revokeCountCap()
+      }
+   }
+}
+```
+
+## Deployment options
+
+Just making a breaking change by removing the linking API and adding the CapCons API might cause some headaches. In order to make the transition to CapCons smother, the CapCons API could temporarily work in parallel with the current linking API.
+
+The rollout process would have 2 steps:
+
+1. Add CapCons and still have linking.
+2. After a transition period remove linking.
+
+To make this FLIP compatible with this rollout, the requirement to removing the private domain needs to be addressed While both APIs are still active, the private domain still needs to exist in order to do private linking for private capabilities. The private domain can be removed at step 2.
+
+During step 1. all existing links should also become CapCons (most likely with a storage migration). The linking API will still be available, but will actually operate with CapCons in the background. Below is a implementation of the linking API with the CapCons API.
+
+```cadence
+// link
+// issuer.link<T>(publicOrPrivatePath, target: storagePath)
+
+let cap = issuer.issueCapability<T>(storagePath)
+issuer.save(cap, to: publicOrPrivatePath)
+```
+
+```cadence 
+// unlink
+// issuer.unlink(publicOrPrivatePath)
+
+let cap = issuer.getCapability<T>(publicOrPrivatePath)
+let capabilityID = cap.capabilityID
+let capCon = issuer.getController(capabilityID: capabilityID)
+
+capCon.revoke()
+```
+
+```cadence
+// relink
+// issuer.unlink(publicOrPrivatePath)
+// issuer.link<&{CounterContract.HasCount}>(publicOrPrivatePath, target: storagePath2)
+
+// 1. unlink
+
+let cap = issuer.getCapability<T>(publicOrPrivatePath)
+let capabilityID = cap.capabilityID
+let capCon = issuer.getController(capabilityID: capabilityID)
+
+capCon.revoke()
+
+// 2. link
+
+let cap = issuer.issueCapability<T>(storagePath2)
+issuer.save(cap, to: publicOrPrivatePath)
+```
+
+
+## Issues not addressed in this FLIP
+
+### Unexpected revocation
+
+If the issuer creates a capability **A** and gives it to Alice and creates a capability **B** and gives it to Bob. Alice then copies her capability creating **A'** and gives it to Charlie, creating the following situation:
+
+- Alice: has **A**
+- Bob: has **B**
+- Charlie: has **A'**
+
+If the issuer decides to revoke **B**, Bob will no longer be able to use his capability. If the issuer revokes **A** Alice will not be able to use her capability, but perhaps unexpected to Charlie, he will also not be able to use **A'**.
+
+Addressing this issue so that it would be clearer to Charlie that he received a capability that is a copy of Alice's, and not his own instance is not in the scope of this FLIP. 
```

Comment:
```suggestion
Addressing this issue so that it would be clearer to Charlie that he received a capability that is a copy of Alice's, and not his own instance, is not in the scope of this FLIP. 
```